File: backend/src/main/java/pers/eastwind/billmanager/reimburse/service/ReimburseService.java
package pers.eastwind.billmanager.reimburse.service;

import jakarta.persistence.criteria.Predicate;
import lombok.extern.slf4j.Slf4j;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.PageRequest;
import org.springframework.data.domain.Pageable;
import org.springframework.data.domain.Sort;
import org.springframework.data.jpa.domain.Specification;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;
import org.springframework.transaction.support.TransactionTemplate;
import pers.eastwind.billmanager.attach.model.Attachment;
import pers.eastwind.billmanager.attach.model.BillType;
import pers.eastwind.billmanager.attach.model.FileOp;
import pers.eastwind.billmanager.attach.model.FileOpType;
import pers.eastwind.billmanager.attach.service.AttachmentService;
import pers.eastwind.billmanager.attach.service.OfficeFileService;
import pers.eastwind.billmanager.attach.util.FileTxUtil;
import pers.eastwind.billmanager.attach.util.FileUtil;
import pers.eastwind.billmanager.common.exception.BizException;
import pers.eastwind.billmanager.common.model.ActionsResult;
import pers.eastwind.billmanager.reimburse.model.*;
import pers.eastwind.billmanager.reimburse.repository.ReimburseRepository;

import java.math.BigDecimal;
import java.nio.file.Files;
import java.nio.file.Path;
import java.security.SecureRandom;
import java.time.Instant;
import java.time.ZoneId;
import java.time.format.DateTimeFormatter;
import java.util.ArrayList;
import java.util.List;
import java.util.stream.Collectors;

/**
 * 报销单服务
 */
@Slf4j
@Service
public class ReimburseService {
    private final ReimburseMapper reimburseMapper;
    private final AttachmentService attachmentService;
    private final ReimburseRepository reimburseRepository;
    private final TransactionTemplate transactionTemplate;
    private final OfficeFileService officeFileService;

    public ReimburseService(ReimburseMapper reimburseMapper1, AttachmentService attachmentService, ReimburseRepository reimburseRepository, TransactionTemplate transactionTemplate, OfficeFileService officeFileService) {
        this.reimburseMapper = reimburseMapper1;
        this.attachmentService = attachmentService;
        this.reimburseRepository = reimburseRepository;
        this.transactionTemplate = transactionTemplate;
        this.officeFileService = officeFileService;
    }

    /**
     * 根据 id 查询
     *
     * @param id ID
     * @return ReimbursementDTO
     */
    public ReimbursementDTO findById(Integer id) {
        if (id == null) {
            throw new BizException("id不能为空");
        }
        Reimbursement bill = reimburseRepository.findById(id).orElse(null);
        if (bill == null) {
            throw new BizException("单据不存在");
        }
        List<Attachment> attachments = attachmentService.getByBill(id, BillType.REIMBURSEMENT);
        return reimburseMapper.toDTO(bill, attachments);
    }

    /**
     * 创建单据
     *
     * @param reimbursementDTO 单据
     * @return 保存后的单据
     */
    @Transactional
    public ReimbursementDTO create(ReimbursementDTO reimbursementDTO) {
        if (reimbursementDTO.getId() != null && reimburseRepository.existsById(reimbursementDTO.getId())) {
            throw new BizException("单据已存在");
        }
        if (reimbursementDTO.getNumber() == null || reimbursementDTO.getNumber().isEmpty()) {
            reimbursementDTO.setNumber("R" + DateTimeFormatter.ofPattern("yyyyMMdd").withZone(ZoneId.systemDefault()).format(Instant.now()) + new SecureRandom().nextInt(1000));
        } else {
            if (reimburseRepository.existsByNumber(reimbursementDTO.getNumber())) {
                throw new BizException("单据编号已存在");
            }
        }
        Reimbursement bill = reimburseRepository.save(reimburseMapper.toEntity(reimbursementDTO));
        attachmentService.updateRelativeAttach(bill.getId(), bill.getNumber(), BillType.REIMBURSEMENT, reimbursementDTO.getAttachments());
        return reimburseMapper.toDTO(bill, attachmentService.getByBill(bill.getId(), BillType.REIMBURSEMENT));
    }


    /**
     * 更新单据
     *
     * @param reimbursementDTO 单据
     * @return 更新后的单据
     */
    @Transactional
    public ReimbursementDTO update(ReimbursementDTO reimbursementDTO) {
        if (reimbursementDTO.getId() == null) {
            throw new BizException("id不能为空");
        }
        Reimbursement bill = reimburseRepository.findById(reimbursementDTO.getId()).orElse(null);
        if (bill == null) {
            throw new BizException("单据不存在");
        }
        reimburseMapper.updateEntityFromDTO(reimbursementDTO, bill);
        bill = reimburseRepository.save(reimburseMapper.toEntity(reimbursementDTO));
        attachmentService.updateRelativeAttach(bill.getId(), bill.getNumber(), BillType.REIMBURSEMENT, reimbursementDTO.getAttachments());
        return reimburseMapper.toDTO(bill, attachmentService.getByBill(bill.getId(), BillType.REIMBURSEMENT));
    }

    /**
     * 根据条件查询
     *
     * @param param 查询参数
     * @return 分页结果
     */
    public Page<ReimbursementDTO> findByParam(ReimburseQueryParam param) {
        if (param == null) {
            throw new BizException("查询参数为空");
        }
        Specification<Reimbursement> specification = (root, query, cb) -> {
            List<Predicate> predicates = new ArrayList<>();
            if (param.getNumber() != null) {
                predicates.add(cb.like(root.get("number"), param.getNumber() + "%"));
            }
            if (param.getSummary() != null) {
                predicates.add(cb.like(root.get("summary"), "%" + param.getSummary() + "%"));
            }
            if (param.getStates() != null && !param.getStates().isEmpty()) {
                if (param.getStates().size() == 1) {
                    predicates.add(cb.equal(root.get("state"), param.getStates().getFirst()));
                } else {
                    predicates.add(root.get("state").in(param.getStates()));
                }
            }
            if (param.getReimburseStartDate() != null) {
                predicates.add(cb.greaterThanOrEqualTo(root.get("reimburseDate"), param.getReimburseStartDate()));
            }
            if (param.getReimburseEndDate() != null) {
                predicates.add(cb.lessThanOrEqualTo(root.get("reimburseDate"), param.getReimburseEndDate()));
            }
            return cb.and(predicates.toArray(new Predicate[0]));
        };
        // 默认取前 20 行，创建时间降序排序
        int pageIndex = param.getPageIndex() == null ? 0 : param.getPageIndex();
        int pageSize = param.getPageSize() == null ? 20 : param.getPageSize();
        List<Sort.Order> orders = param.getSorts() == null ? List.of(Sort.Order.desc("reimburseDate")) : param.getSorts().stream().map(sortParam -> Sort.Order.by(sortParam.getField()).with(Sort.Direction.fromString(sortParam.getDirection()))).toList();
        Pageable pageable = PageRequest.of(pageIndex, pageSize, Sort.by(orders));
        Page<Reimbursement> pageResult = reimburseRepository.findAll(specification, pageable);
        return pageResult.map(reimburseMapper::toBaseDTO);
    }

    /**
     * 批量删除单据
     *
     * @param ids 单据 ID 列表
     * @return 批量操作结果
     */
    public ActionsResult<Integer, Void> delete(List<Integer> ids) {
        if (ids == null || ids.isEmpty()) {
            throw new BizException("id不能为空");
        }
        return ActionsResult.executeActions(ids, id -> {
            transactionTemplate.executeWithoutResult(status -> {
                Reimbursement bill = reimburseRepository.findById(id).orElse(null);
                if (bill == null) {
                    throw new BizException("单据不存在");
                }
                if (bill.getState() != ReimburseState.CREATED) {
                    throw new BizException("非创建状态不能删除");
                }
                reimburseRepository.deleteById(id);
                attachmentService.updateRelativeAttach(bill.getId(), bill.getNumber(), BillType.REIMBURSEMENT, new ArrayList<>());
            });
            return null;
        });
    }

    /**
     * 提交处理
     */
    public ActionsResult<Integer, Void> process(List<Integer> ids) {
        if (ids == null || ids.isEmpty()) {
            throw new BizException("id不能为空");
        }
        return ActionsResult.executeActions(ids, id -> {
            transactionTemplate.executeWithoutResult(status -> {
                Reimbursement bill = reimburseRepository.findById(id).orElse(null);
                if (bill == null) {
                    throw new BizException("单据不存在");
                }
                if (bill.getState() != ReimburseState.CREATED) {
                    throw new BizException("非创建状态不能提交");
                }
                bill.setState(ReimburseState.PROCESSING);
                reimburseRepository.save(bill);
            });
            return null;
        });
    }

    /**
     * 完成
     */
    public ActionsResult<Integer, Void> finish(List<Integer> ids) {
        if (ids == null || ids.isEmpty()) {
            throw new BizException("id不能为空");
        }
        return ActionsResult.executeActions(ids, id -> {
            transactionTemplate.executeWithoutResult(status -> {
                Reimbursement bill = reimburseRepository.findById(id).orElse(null);
                if (bill == null) {
                    throw new BizException("单据不存在");
                }
                if (bill.getState() != ReimburseState.PROCESSING) {
                    throw new BizException("非处理状态不能完成");
                }
                bill.setState(ReimburseState.FINISHED);
                reimburseRepository.save(bill);
            });
            return null;
        });
    }

    /**
     * 导出单据
     *
     * @param ids 单据列表
     * @return 压缩文件路径
     */
    public Path export(List<Integer> ids) {
        if (ids == null || ids.isEmpty()) {
            throw new BizException("id不能为空");
        }
        List<Reimbursement> reimbursements = reimburseRepository.findAllById(ids);
        if (reimbursements.isEmpty()) {
            throw new BizException("id不存在");
        }
        // 临时目录
        Path tempPath = attachmentService.createTempDir("export");
        // 文件操作
        List<FileOp> ops = new ArrayList<>();

        // 遍历生成 excel行，并拷贝附件
        List<List<String>> rows = new ArrayList<>();
        rows.add(List.of("单据编号", "摘要", "总额", "报销日期", "备注"));
        BigDecimal totalAmount = BigDecimal.ZERO;
        DateTimeFormatter dateTimeFormatter = DateTimeFormatter.ofPattern("yyyy-MM-dd");
        for (Reimbursement reimbursement : reimbursements) {
            rows.add(List.of(
                    reimbursement.getNumber(),
                    reimbursement.getSummary(),
                    reimbursement.getTotalAmount().toString(),
                    reimbursement.getReimburseDate() == null ? "" : dateTimeFormatter.format(reimbursement.getReimburseDate().atZone(ZoneId.systemDefault())),
                    reimbursement.getDetails().stream().map((detail) ->
                                    detail.getName() + " : " + detail.getAmount().stripTrailingZeros().toPlainString() + " ; ")
                            .collect(Collectors.joining())
            ));
            totalAmount = totalAmount.add(reimbursement.getTotalAmount());
            // 创建当前单据附件文件夹
            Path curDir = tempPath.resolve(reimbursement.getNumber());
            // 拷贝当前单据所有附件
            List<Attachment> attachments = attachmentService.getByBill(reimbursement.getId(), BillType.REIMBURSEMENT);
            for (Attachment attachment : attachments) {
                Path origin = attachmentService.getRootPath().resolve(attachment.getRelativePath());
                Path target = curDir.resolve(attachment.getName());
                // 处理可能的重名
                int repeatCount = 1;
                while (Files.exists(target)) {
                    target = curDir.resolve(repeatCount + "-" + attachment.getName());
                    repeatCount++;
                }
                ops.add(new FileOp(FileOpType.COPY, origin, target));
            }
        }
        // 表合计
        rows.add(List.of("", "合计", totalAmount.toString(), "", ""));

        Path excel = tempPath.resolve("导出结果.xlsx");
        officeFileService.generateExcelFromList(rows, excel);
        // 执行文件拷贝
        FileTxUtil.exec(ops);
        // 生成压缩包
        Path zip = attachmentService.getTempPath().resolve(tempPath + ".zip");
        FileUtil.zip(tempPath, zip);
        return zip;
    }
}

File: backend-go/internal/reimburse/service/reimburse_service.go
package service

import (
	attachModel "backend-go/internal/attach/model"
	"backend-go/internal/common/cache"
	"context"
	"fmt"
	"math/rand"
	"path/filepath"
	"time"

	"backend-go/internal/attach/files"
	"backend-go/internal/attach/service"
	"backend-go/internal/common/errs"
	"backend-go/internal/common/result"
	"backend-go/internal/reimburse/model"
	"backend-go/internal/reimburse/repository"
)

type ReimburseService struct {
	reimburseRepo   *repository.ReimburseRepository
	attachService   *service.AttachmentService
	cache           cache.Cache
	reimburseMapper *ReimburseMapper
}

type ReimburseMapper struct{}

func NewReimburseService(
	reimburseRepo *repository.ReimburseRepository,
	attachService *service.AttachmentService,
	cache cache.Cache,
) *ReimburseService {
	return &ReimburseService{
		reimburseRepo: reimburseRepo,
		attachService: attachService,
		cache:         cache,
	}
}

func (s *ReimburseService) GenerateNumber() string {
	timestamp := time.Now().Format("20060102")
	randomNum := rand.Intn(1000)
	return fmt.Sprintf("R%s%04d", timestamp, randomNum)
}

func (s *ReimburseService) FindByID(ctx context.Context, id uint) (*model.ReimbursementDTO, error) {
	if id == 0 {
		return nil, errs.NewBizError("ID不能为空")
	}
	bill, err := s.reimburseRepo.FindFullById(ctx, id)
	if err != nil {
		return nil, err
	}
	if bill == nil {
		return nil, errs.NewBizError("单据不存在")
	}
	attachments, err := s.attachService.GetByBill(ctx, bill.ID, attachModel.BillTypeReimbursement)
	if err != nil {
		return nil, errs.NewBizError("获取附件失败")
	}
	return bill.ToDTO(attachments), nil
}

func (s *ReimburseService) Create(ctx context.Context, dto *model.ReimbursementDTO) (*model.ReimbursementDTO, error) {
	if dto.ID != 0 {
		dto.ID = 0
	}
	entity := model.Reimbursement{
		ID:            dto.ID,
		Number:        dto.Number,
		State:         dto.State,
		Summary:       dto.Summary,
		TotalAmount:   dto.TotalAmount,
		ReimburseDate: dto.ReimburseDate,
		Remark:        dto.Remark,
	}
	if entity.Number == "" {
		entity.Number = s.GenerateNumber()
	}
	err := s.reimburseRepo.Transaction(ctx, func(tx context.Context) error {
		exists, _ := s.reimburseRepo.ExistsByNumber(ctx, dto.Number)
		if exists {
			return errs.NewBizError("单据编号已存在")
		}
		if err := s.reimburseRepo.Create(tx, &entity); err != nil {
			return err
		}

		if err := s.attachService.UpdateRelativeAttach(tx, entity.ID, entity.Number, attachModel.BillTypeReimbursement, &dto.Attachments); err != nil {
			return err
		}
		return nil
	})

	if err != nil {
		return nil, err
	}

	attaches, err := s.attachService.GetByBill(ctx, entity.ID, attachModel.BillTypeReimbursement)
	if err != nil {
		return nil, err
	}
	return entity.ToDTO(attaches), nil
}

func (s *ReimburseService) Update(ctx context.Context, dto *model.ReimbursementDTO) (*model.ReimbursementDTO, error) {
	if dto.ID == 0 {
		return nil, errs.NewBizError("id不能为空")
	}

	entity := model.Reimbursement{
		ID:            dto.ID,
		Number:        dto.Number,
		State:         dto.State,
		Summary:       dto.Summary,
		TotalAmount:   dto.TotalAmount,
		ReimburseDate: dto.ReimburseDate,
		Remark:        dto.Remark,
	}

	if len(dto.Details) > 0 {
		details := make([]model.ReimburseDetail, len(dto.Details))
		for i, detail := range dto.Details {
			details[i] = model.ReimburseDetail{
				ID:     detail.ID,
				Name:   detail.Name,
				Amount: detail.Amount,
			}
		}
		entity.Details = details
	}

	err := s.reimburseRepo.Transaction(ctx, func(tx context.Context) error {
		bill, err := s.reimburseRepo.FindByID(tx, dto.ID)
		if err != nil {
			return err
		}
		if bill == nil {
			return errs.NewBizError("单据不存在")
		}
		err = s.attachService.UpdateRelativeAttach(tx, entity.ID, entity.Number, attachModel.BillTypeReimbursement, &dto.Attachments)
		if err != nil {
			return err
		}
		return s.reimburseRepo.Updates(tx, bill)
	})

	if err != nil {
		return nil, err
	}

	attaches, err := s.attachService.GetByBill(ctx, entity.ID, attachModel.BillTypeReimbursement)
	if err != nil {
		return nil, err
	}
	return entity.ToDTO(attaches), nil
}

func (s *ReimburseService) FindByParam(ctx context.Context, param *model.ReimburseQueryParam) (*result.PageResult[model.ReimbursementDTO], error) {
	if param == nil {
		return nil, errs.NewBizError("查询参数为空")
	}

	res, err := s.reimburseRepo.FindByParam(ctx, param)
	if err != nil {
		return nil, err
	}

	return result.NewPageResultFromDB(res, model.ToBaseDTOs), nil
}

func (s *ReimburseService) Delete(ctx context.Context, ids []uint) (*result.ActionsResult[uint, any], error) {
	if len(ids) == 0 {
		return nil, errs.NewBizError("ID 为空")
	}
	return result.ExecuteActions(ids, func(id uint) (any, error) {
		err := s.reimburseRepo.Transaction(ctx, func(tx context.Context) error {
			bill, err := s.reimburseRepo.FindByID(tx, id)
			if err != nil {
				return err
			}
			if bill == nil {
				return errs.NewBizError("单据不存在")
			}
			if bill.State != model.ReimburseStateCreated {
				return errs.NewBizError("非创建状态不能删除")
			}
			if err = s.reimburseRepo.DeleteByID(tx, id); err != nil {
				return err
			}
			return s.attachService.UpdateRelativeAttach(tx, bill.ID, bill.Number, attachModel.BillTypeReimbursement, nil)
		})
		return nil, err
	}), nil
}

func (s *ReimburseService) Process(ctx context.Context, ids []uint) (*result.ActionsResult[uint, any], error) {
	if len(ids) == 0 {
		return nil, errs.NewBizError("ID 为空")
	}
	return result.ExecuteActions(ids, func(id uint) (any, error) {
		err := s.reimburseRepo.Transaction(ctx, func(tx context.Context) error {
			bill, err := s.reimburseRepo.FindByID(tx, id)
			if err != nil {
				return err
			}
			if bill == nil {
				return errs.NewBizError("单据不存在")
			}
			if bill.State != model.ReimburseStateCreated {
				return errs.NewBizError("非创建状态不能提交")
			}
			bill.State = model.ReimburseStateProcessing
			return s.reimburseRepo.Updates(tx, bill)
		})
		return nil, err
	}), nil
}

func (s *ReimburseService) Finish(ctx context.Context, ids []uint) (*result.ActionsResult[uint, any], error) {
	if len(ids) == 0 {
		return nil, errs.NewBizError("ID 为空")
	}
	return result.ExecuteActions(ids, func(id uint) (any, error) {
		err := s.reimburseRepo.Transaction(ctx, func(tx context.Context) error {
			bill, err := s.reimburseRepo.FindByID(tx, id)
			if err != nil {
				return err
			}
			if bill == nil {
				return errs.NewBizError("单据不存在")
			}
			if bill.State != model.ReimburseStateProcessing {
				return errs.NewBizError("非处理状态不能完成")
			}
			bill.State = model.ReimburseStateFinished
			return s.reimburseRepo.Updates(tx, bill)
		})
		return nil, err
	}), nil
}

func (s *ReimburseService) Export(ctx context.Context, ids []uint) (string, error) {
	if ids == nil || len(ids) == 0 {
		return "", errs.NewBizError("ID 不能为空")
	}

	reimbursements, err := s.reimburseRepo.FindAll(ctx, ids)
	if reimbursements == nil || len(*reimbursements) == 0 {
		return "", errs.NewBizError("单据不存在")
	}

	// 创建临时目录
	tempDir, err := s.attachService.CreateTempDir(s.cache, "export")
	if err != nil {
		return "", err
	}

	var ops []attachModel.FileOp
	var rows [][]string

	// 表头
	rows = append(rows, []string{"单据编号", "摘要", "总额", "报销日期", "备注"})
	var totalAmount float64

	// 遍历生成 excel行，并拷贝附件
	for _, reimbursement := range *reimbursements {
		dateStr := ""
		if reimbursement.ReimburseDate != nil {
			dateStr = reimbursement.ReimburseDate.Format("2006-01-02")
		}

		// 构建详情字符串
		detailStr := ""
		for _, detail := range reimbursement.Details {
			detailStr += fmt.Sprintf("%s : %.2f ; ", detail.Name, detail.Amount)
		}

		// 添加行数据
		rows = append(rows, []string{
			reimbursement.Number,
			reimbursement.Summary,
			fmt.Sprintf("%.2f", reimbursement.TotalAmount),
			dateStr,
			detailStr,
		})

		totalAmount += reimbursement.TotalAmount

		// 获取当前单据的所有附件
		attachments, err := s.attachService.GetByBill(ctx, reimbursement.ID, attachModel.BillTypeReimbursement)
		if err != nil {
			return "", err
		}

		// 创建当前单据附件文件夹
		curDir := filepath.Join(tempDir, reimbursement.Number)
		for _, attachment := range *attachments {
			// 获取原始附件路径
			originPath, err := s.attachService.GetAbsolutePath(attachment.RelativePath, false)
			if err != nil {
				return "", err
			}

			// 设置目标路径
			targetPath := filepath.Join(curDir, attachment.Name)

			// 处理可能的重名
			repeatCount := 1
			for files.Exists(targetPath) {
				targetPath = fmt.Sprintf("%s/%d-%s", curDir, repeatCount, attachment.Name)
				repeatCount++
			}
			// 添加文件操作到列表
			ops = append(ops, attachModel.FileOp{
				Type:   attachModel.FileOpCopy,
				Origin: originPath,
				Target: targetPath,
			})
		}
	}

	// 表合计
	rows = append(rows, []string{"", "合计", fmt.Sprintf("%.2f", totalAmount), "", ""})

	// 生成Excel文件
	excelPath := filepath.Join(tempDir, fmt.Sprintf("导出结果%s.xlsx", time.Now().Format("20060102150405")))
	if err = files.GenerateExcelFromList(&rows, excelPath); err != nil {
		return "", errs.NewBizError("生成Excel失败: " + err.Error())
	}

	if err = files.Exec(s.cache, &ops); err != nil {
		return "", errs.NewBizError("文件操作失败: " + err.Error())
	}
	zipPath, err := files.Zip(tempDir, "")
	if err != nil {
		return "", err
	}
	return zipPath, nil
}
